### DISCUSSION TITLE
Discrepancy in Layer and Parameter Counts During RT-DETR-L Fine-tuning

### DISCUSSION THREAD

**Discussion by nitkizs** [ultralytics/ultralytics#22075] [Q&A]
Hi team,

I'm working with the RT-DETR models from Ultralytics and have observed a behavior during fine-tuning that I'd like to understand better.

I ran model.info() on two different versions of the RT-DETR-L model (the official rtdetr-l.pt and a scaled-down version) before and after fine-tuning them on a 6-class custom dataset.

- Observation 1: Original rtdetr-l.pt

  - Pretrained: 449 layers, 32,970,476 parameters.

  - Fine-tuned: 457 layers, 32,818,406 parameters.

1. Question: Why did the layer count increase from 449 to 457 after fine-tuning, in addition to the expected change in parameters from resizing the output heads? Does the fine-tuning script add non-trainable modules like ReLU that the summary function counts as layers in this specific model version?

- Observation 2: Scaled-down rtdetr-l (Scaling parameters used: depth=0.5, width=0.5, max_channels=1024)

  - Pretrained: 361 layers, 19,327,052 parameters.

  - Fine-tuned: 361 layers, 19,174,982 parameters.

2. Question: In this case, the layer count remains constant, while the parameter count decreases, as expected from resizing the final classification heads. Why does the layer count remain the same here, while it increased in Observation 1, assuming I am using the same version of the Ultralytics framework and the same model.info() method?

Is the difference in behavior related to an architectural variation between the original and scaled-down rtdetr-l models ? Could you provide insight into the specific fine-tuning recipe or architectural differences between these two model versions that would explain this discrepancy in layer counts?

Thank you for your help

**Comment by UltralyticsAssistant** [comment]
👋 Hello @nitkizs, thank you for your detailed report and for using Ultralytics 🚀! This is an automated response to help you get support quickly. An Ultralytics engineer will also review and assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples and where many common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us investigate the layer/parameter count discrepancy you observed. When possible, include:
-Exact ultralytics version and install method (output of `pip show ultralytics`)
-Exact PyTorch, CUDA, CUDNN, Python versions, OS, and device info
-The exact model files used (e.g., rtdetr-l.pt source) and any custom YAML or scaling args (depth/width/max_channels)
-Your training/finetuning command(s), seeds, and logs
-Full `model.info()` outputs before and after fine-tuning for both models
-The dataset YAML (with nc=6) and any relevant snippets
-A minimal script or notebook showing the steps to reproduce

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Screenshots or small code snippets are great 👍

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
The pretrained model does not have activation as a layer (`self.act`) in the head. However, it's used in the `forward` pass:
https://github.com/ultralytics/ultralytics/blob/dbc5433bc8058e28ecea422c7c2210a9acae7a32/ultralytics/nn/modules/transformer.py#L408

The fine-tuned model adds ReLU activation as a layer so that it doesn't have to generate it repeatedly in the forward pass.
